File: Learning/src/Questions/BinaryTree/PreOrderTraversal.java
package Questions.BinaryTree;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class PreOrderTraversal {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }

    }
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Stack<TreeNode> st = new Stack<>();

        while (root != null || !st.isEmpty()) {
            while (root != null) {
                result.add(root.val);  // Process the current node
                st.push(root);
                root = root.left;
            }

            root = st.pop();
            root = root.right;
        }

        return result;
    }
    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        TreeNode right = new TreeNode(2);
        TreeNode rightLeft = new TreeNode(3);
        root.right = right;
        right.left = rightLeft;
        PreOrderTraversal preOrderTracersal = new PreOrderTraversal();
        List<Integer> result = preOrderTracersal.preorderTraversal(root);
        System.out.println("PreOrder Traversal: " + result);
    }
}

File: DSA/Python/Questions/BinaryTree/PreOrderTraversal.py
"""
# Question: Binary Tree Preorder Traversal
# Link: https://leetcode.com/problems/binary-tree-preorder-traversal/

# Time Complexity: O(N)
# Space Complexity: O(H)

# Algorithm:
# 1. Root -> Left -> Right traversal
# 2. Iterative and recursive solutions
# 3. Morris traversal option
"""


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class PreOrderTraversal:
    def preorder_recursive(self, root: TreeNode) -> list[int]:
        result = []

        def dfs(node: TreeNode) -> None:
            if node:
                result.append(node.val)
                dfs(node.left)
                dfs(node.right)

        dfs(root)
        return result

    def preorder_iterative(self, root: TreeNode) -> list[int]:
        if not root:
            return []

        result = []
        stack = [root]

        while stack:
            node = stack.pop()
            result.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)

        return result


def main():
    solution = PreOrderTraversal()
    root = TreeNode(1)
    root.right = TreeNode(2)
    root.right.left = TreeNode(3)
    print(solution.preorder_recursive(root))
    print(solution.preorder_iterative(root))


if __name__ == "__main__":
    main()
